File: day5/js/main.js
const fs = require('fs');
const input = readInput(fs.readFileSync('../test.txt', 'utf8'));
showResult("part one = ", partOne, input);
showResult("part two = ", partTwo, input);

function showResult(str, func, arg) {
    const start = performance.now();
    const res = func(arg)
    const end = performance.now();
    console.log(str, res);
    console.log("time[ms] = ", end - start);
    return res;
}

function readInput(input) {
    const seeds = [];
    const lines = input.split('\n');
    for (let numstr of lines[0].substring(lines[0].indexOf(':') + 2).split(' ')) {
        seeds.push(Number(numstr));
    }

    let map = [];
    const maps = [];
    for (let line of lines.splice(3)) {
        if (!isNaN(line[0]) && line[0] != '\r') {
            const splits = line.split(' ');
            const dst = Number(splits[0]);
            const src = Number(splits[1]);
            const len = Number(splits[2]);
            map.push({lo: src, hi: src + len - 1, slide: dst - src});
        }
        else if (map.length) {
            maps.push(map);
            map = [];
        }
    }
    maps.push(map);
    return { seeds, maps };
}

function partOne(input) {
    const mins = [];
    for (let i = 0; i < input.seeds.length; i += 1) {
        mins.push(sandFilter(input.maps, { lo: input.seeds[i], hi: input.seeds[i] }));
    }
    return Math.min(...mins);
}

function partTwo(input) {
    const mins = [];
    for (let i = 0; i < input.seeds.length; i += 2) {
        mins.push(sandFilter(input.maps, { lo: input.seeds[i], hi: input.seeds[i] + input.seeds[i + 1] - 1 }));
    }
    return Math.min(...mins);
}

function sandFilter(maps, range) {
    let ranges = [];
    ranges.push(range);
    for (map of maps) {
        const newRanges = [];
        while (ranges.length) {
            const range = ranges.pop();
            let passed = false;
            for (xform of map) {
                if ((range.lo < xform.lo) && (range.hi < xform.lo)) {
                    passed = true;
                }
                else if ((range.lo > xform.hi) && (range.hi > xform.hi)) {
                    passed = true;
                }
                else if ((range.lo < xform.lo) && ( range.hi >= xform.lo)) {
                    ranges.push({lo: range.lo, hi: xform.lo - 1});
                    newRanges.push({lo: xform.lo + xform.slide, hi: range.hi + xform.slide});
                    passed = false;
                    break;
                }
                else if ((range.lo >= xform.lo) && (range.hi > xform.hi)) {
                    ranges.push({lo: xform.hi + 1, hi: range.hi});
                    newRanges.push({lo: range.lo + xform.slide, hi: xform.hi + xform.slide});
                    passed = false;
                    break;
                }
                else {
                    newRanges.push({lo: range.lo + xform.slide, hi: range.hi + xform.slide});
                    passed = false;
                    break;
                }
            }
            if (passed) {
                newRanges.push(range);
            }
        }
        ranges = newRanges
    }
    let min = Infinity;
    ranges.map((x) => {
        if (x.lo < min) min = x.lo;
    });
    return min;
}

File: day5/cs/main.cs
using System.Diagnostics;
using Map = System.Collections.Generic.List<Xform>;

Input readInput(string fileName)
{
    List<Map> maps = new List<Map>();
    StreamReader sr = new StreamReader(fileName);
    string line;

    line = sr.ReadLine();
    long[] seeds = line.Replace("seeds: ", "").Split(' ').Select(x => long.Parse(x)).ToArray();

    Map map = new Map(); 
    while ((line = sr.ReadLine()) != null)
    {
        if (line.Length == 0) continue;
        if (Char.IsLetter(line[0]))
        {
            if (map.Count > 0)
            {
                maps.Add(map);
                map = new Map();
                continue;
            }
        } 
        else
        {
            long[] nums = line.Split(' ').Select(x => long.Parse(x)).ToArray();
            long dst = nums[0];
            long src = nums[1];
            long len = nums[2];
            map.Add(new Xform() { start = src, end = src + len - 1, slide = dst - src });
        }
    }
    maps.Add(map);

    return new Input() { seeds = seeds, maps = maps };
}

long filterSand(List<Map> maps, Range _range)
{
    Stack<Range> ranges = new Stack<Range>();
    ranges.Push(_range);
    foreach (Map map in maps)
    {
        Stack<Range> newRanges = new Stack<Range>(32);
        while (ranges.Count > 0)
        {
            Range range = ranges.Pop();
            long rs = range.start;
            long re = range.end;
            bool passed = true;
            foreach (Xform xform in map)
            {
                long ms = xform.start;
                long me = xform.end;
                long s = xform.slide;
                if ((rs < ms) && (re < ms)) {}
                else if ((rs > me) && (re > me)) {}
                else if ((rs < ms) && (re >= ms))
                {
                    ranges.Push(new Range() { start = rs, end = ms - 1 });
                    newRanges.Push(new Range() { start = ms + s, end = re + s });
                    passed = false;
                    break;
                }
                else if ((rs >= ms) && (re > me))
                {
                    ranges.Push(new Range() { start = me + 1, end = re });
                    newRanges.Push(new Range() { start = rs + s, end = me + s });
                    passed = false;
                    break;
                }
                else
                {
                    newRanges.Push(new Range() { start = rs + s, end = re + s });
                    passed = false;
                    break;
                }
            }
            if (passed)
            {
                newRanges.Push(range);
            }
        }
        ranges = newRanges;
    }

    List<Range> sortedRanges = ranges.ToList();
    sortedRanges.Sort(delegate (Range r1, Range r2)
    {
        return (r1.start < r2.start) ? -1 : 1;
    });
    return sortedRanges[0].start;
}
long filterSandSimple(List<Map> maps, long seed)
{
    foreach (Map map in maps)
    {
        foreach (Xform xform in map)
        {
            if (seed >= xform.start && seed <= xform.end)
            {
                seed += xform.slide;
                break;
            }
            else { continue; }
            break;
        }
    }
    return seed;
}
long partOne(Input input)
{
    long min = long.MaxValue;
    foreach (long seed in input.seeds)
    {
        long res = filterSandSimple(input.maps, seed);
        if (res < min)
        {
            min = res;
        }
    }
    return min;
}

long partTwo(Input input)
{
    long min = long.MaxValue;
    for (int i = 0; i < input.seeds.Length; i += 2)
    {
        long res = filterSand(input.maps, new Range() { 
            start = input.seeds[i], end = input.seeds[i] + input.seeds[i + 1] - 1 });
        if (res < min)
        {
            min = res;
        }
    }
    return min;
}

void show<T1, T2>(string s, Func<T1, T2> func, T1 arg)
{
    Stopwatch timer = new Stopwatch();
    timer.Start();
    T2 res = func(arg);
    timer.Stop();
    s += "{0}";
    Console.WriteLine(String.Format(s, res));
    Console.WriteLine(String.Format("time[us] = {0}", timer.Elapsed.TotalMicroseconds));
}
T2 measure<T1, T2>(string s, Func<T1, T2> func, T1 arg)
{
    Stopwatch timer = new Stopwatch();
    timer.Start();
    T2 res = func(arg);
    timer.Stop();
    Console.WriteLine(s);
    Console.WriteLine(String.Format("time[us] = {0}", timer.Elapsed.TotalMicroseconds));
    return res;
}

Input input = measure<string, Input>("loading", readInput, @"..\..\..\..\..\input.txt");
show<Input, long>("part one = ", partOne, input);
show<Input, long>("part two = ", partTwo, input);


struct Range
{
    public long start, end;
}
struct Xform
{
    public long start, end, slide;
}
struct Input
{
    public long[] seeds;
    public List<Map> maps;
}
